### DISCUSSION TITLE
Deconstruct properties

### DISCUSSION THREAD

**Discussion by ziaulhasanhamim** [dotnet/csharplang#6355] [Language Ideas] [answered]
C# already has support for deconstructing objects. 

```C#
A a = new();
if (a is {I: var i})
{
  Console.WriteLine(i);
}

public class A 
{
    public int I { get;set; }
}
```
Isn't it redundant? Because of this kind of property Deconstructing if-check will always succeed.
So why can't we just do this,
```c#
A a = new();
{ I: var i } = a;
Console.WriteLine(i);

// or 
A a = new();
A { I: var i } = a;
Console.WriteLine(i);
```
Note that there will be no runtime pattern matching there. So it can be decided in compile time whether or not the object can be deconstructed with specified properties.
So you can't do this,
```c#
B b = new();
{ I: var i, J: 5 } = b;
Console.WriteLine(i);

public class B
{
    public int I { get;set; }
    public int J { get;set; }
}
```
Because the compiler can't check if The `b.J` value is 5 in compile time.

**Comment by HaloFour** [comment, marked as answer]
See: #4082
